trade-finance-manager-ui/server/controllers/utilisation-reports/index.ts

```diff
@@ -1,30 +1,130 @@
-    const reportPeriod = getFormattedReportPeriod();
-    const reportDueDate = await getFormattedReportDueDate(userToken);
+    const { reportReconciliationSummary, reportPeriod, reportDueDate, month, year } = await getUtilisationReportsSummaryData(userToken);
 
     return res.render('utilisation-reports/utilisation-reports.njk', {
       user,
       activePrimaryNavigation: 'utilisation reports',
       reportReconciliationSummary,
       reportPeriod,
       reportDueDate,
+      month,
+      year,
     });
   } catch (error) {
     console.error('Error rendering utilisation reports page', error);
     return res.render('_partials/problem-with-service.njk', { user });
   }
 };
+
+const getReportIdentifiersFromUrl = (body: object, month: number, year: number): ReportIdentifier[] =>
+  Object.keys(body).reduce((reportIdentifiers, key) => {
+    const match = key.match(/(set-status--)(.*)(-)([a-f\d]+)/);
+    if (!match) {
+      return reportIdentifiers;
+    }
+
+    const idType = match.at(2);
+    const id = match.at(4);
+
+    if (!idType || !id) {
+      throw new Error('Failed to parse report identifiers from request body');
+    }
+
+    if (idType === 'reportId') {
+      return [...reportIdentifiers, { id }];
+    }
+
+    if (idType === 'bankId') {
+      return [...reportIdentifiers, { bankId: id, month, year }];
+    }
+
+    throw new Error('Failed to parse report identifiers from request body');
+  }, [] as ReportIdentifier[]);
+
+export type UpdateReportStatusPayload = {
+  reportsWithStatus: ReportWithStatus[];
+  user: TfmSessionUser;
+};
+
+export const updateUtilisationReportStatus = async (req: Request, res: Response) => {
+  const { user, userToken } = req.session;
+  const { status, month, year } = req.query;
+
+  try {
+    if (!user) {
+      throw new Error('User is not logged in');
+    }
+
+    const reportIdentifiers = getReportIdentifiersFromUrl(req.body as object, Number(month), Number(year));
+
+    if (status !== 'completed' && status !== 'not-completed') {
+      throw new Error('Supplied status instruction does not match any expected status');
+    }
+    const statusToSet =
+      status === 'completed' ? UTILISATION_REPORT_RECONCILIATION_STATUS.RECONCILIATION_COMPLETED : UTILISATION_REPORT_RECONCILIATION_STATUS.REPORT_NOT_RECEIVED;
+
+    const reportsWithStatus = reportIdentifiers.map((reportIdentifier): ReportWithStatus => {
+      if ('id' in reportIdentifier && statusToSet === UTILISATION_REPORT_RECONCILIATION_STATUS.REPORT_NOT_RECEIVED) {
+        return {
+          status: UTILISATION_REPORT_RECONCILIATION_STATUS.PENDING_RECONCILIATION,
+          report: reportIdentifier,
+        };
+      }
+      if ('bankId' in reportIdentifier && statusToSet === UTILISATION_REPORT_RECONCILIATION_STATUS.REPORT_NOT_RECEIVED) {
+        throw new InvalidReportStatusRequestError("Cannot set non-existing report to 'REPORT_NOT_RECEIVED'");
+      }
```

Comment:
I want to discuss this when I'm back, but annoyingly `session.withTransaction` (in `dtfs-central-api`) does not roll back any changes made if an error is thrown inside the function. So, if you try to mark two reports as "not done", one of which is in the `NOT_RECEIVED` state and one in the `RECONCILIATION_COMPLETED` state, an error is thrown and you see the `Problem with service` page. Then, navigating back to "bank reports" shows that the report which was in the `RECONCILIATION_COMPLETED` state is now in the `PENDING_RECONCILIATION` state. To avoid this happening, I catch the attempt to mark a not received report as "not done" in the frontend and show that error on the page (see the screenshot in the PR description)

Reply:
As we don't actually need to handle this logic, we now just filter out any requests where the user tries to mark a non existing report as "not completed"